File: AllSamples/Samples/Samples_Source/natural language/lex/vb/lex.vb
'---------------------------------------------------------------------
'  This file is part of the Microsoft .NET Framework SDK Code Samples.
' 
'  Copyright (C) Microsoft Corporation.  All rights reserved.
' 
'This source code is intended only as a supplement to Microsoft
'Development Tools and/or on-line documentation.  See these other
'materials for detailed information regarding Microsoft code samples.
' 
'THIS CODE AND INFORMATION ARE PROVIDED AS IS WITHOUT WARRANTY OF ANY
'KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
'IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
'PARTICULAR PURPOSE.
'---------------------------------------------------------------------
Imports System
Imports System.IO
Imports System.Collections
Imports System.NaturalLanguage
Imports System.Text
Imports System.Globalization


Namespace Microsoft.Samples.NaturalLanguage
    
   Class LexiconConsole
      

      '/ <summary>
      '/ The main entry point for the Console.
      '/ </summary>
      Overloads Public Shared Function Main(args() As String) As Integer
         Dim lexSample As LexiconExamples = Nothing
         Dim cmdOpts As CmdOpts = Nothing
         Dim outputWriter As StreamWriter = Nothing
         
         Try
            'Process commandline arguments
            cmdOpts = New CmdOpts()
            If Not cmdOpts.ProcessOpts(args) Then
               Throw New ArgumentException("Parsing commandline arguments failed.")
            End If
            
            lexSample = New LexiconExamples()
            If Not (cmdOpts.OutputFile Is Nothing) Then
               outputWriter = New StreamWriter(cmdOpts.OutputFile)
            Else
               outputWriter = New StreamWriter(Console.OpenStandardOutput())
            End If
            
            lexSample.DisplayWords(outputWriter, cmdOpts.InputFile, cmdOpts.Word, cmdOpts.MaxWordsToDisplay)
            Return 0
         Catch nre As System.NullReferenceException
            If cmdOpts.Verbose Then
               Console.Error.WriteLine("Message: {0}", nre.Message)
               Console.Error.WriteLine("Stack Trace: {0}", nre.StackTrace)
            End If
            Return 1
         Catch ae As System.ArgumentException
            If cmdOpts.Verbose Then
               Console.Error.WriteLine("Message: {0}", ae.Message)
               Console.Error.WriteLine("Stack Trace: {0}", ae.StackTrace)
            End If
            
            Return 1
         Finally
            If Not (outputWriter Is Nothing) Then
               outputWriter.Close()
            End If
         End Try
      End Function 'Main
   End Class 'LexiconConsole
    
   Public Class LexiconExamples
      
      Public Sub New()
      End Sub 'New
      
      
      Protected Sub DisplayWordProps(sw As StreamWriter, props As IDictionary, prefix As String)
         Dim pair As DictionaryEntry
         For Each pair In  props
            sw.Write("{0}name: {1}", prefix, pair.Key.ToString())
            
            Dim subProps As IDictionary = Nothing
            Dim subList As ICollection = Nothing
            
            If TypeOf pair.Value Is IDictionary Then
               subProps = CType(pair.Value, IDictionary)
               sw.WriteLine()
               DisplayWordProps(sw, subProps, prefix + "  ")
            Else
               If TypeOf pair.Value Is ICollection Then
                  subList = CType(pair.Value, ICollection)
                        Dim prefix2 As String = prefix & "  "
                  
                  sw.WriteLine()
                  Dim member As Object
                  For Each member In  subList
                     sw.WriteLine("{0} value: {1}", prefix2, member.ToString())
                  Next member
               Else
                  sw.WriteLine(" value: {0}", pair.Value.ToString())
               End If
            End If
         Next pair 
         sw.Flush()
      End Sub 'DisplayWordProps
      
      
      Public Sub DisplayWords(sw As StreamWriter, inputFile As String, word As String, maxWordsToDisplay As Integer)
         Dim lexEnum As LexiconEntryEnumerator = Nothing
         Dim lex As Lexicon = Nothing
         
         If sw Is Nothing Or inputFile Is Nothing Or word Is Nothing Or maxWordsToDisplay <= 0 Then
            Return
         End If 
         Try
            If File.Exists(inputFile) Then
               'Create lexicon based on the lexicon file specified in the commandline
               lex = New Lexicon(inputFile)
               
               'Get a LexiconEntryEnumerator
               lexEnum = CType(lex.GetEnumerator(), LexiconEntryEnumerator)
               
               'Move to the target word if it is specified in the commandline
               If word.Length <> 0 Then
                  Dim id As Integer = lex.IndexOf(word)
                  sw.WriteLine("  IndexOf({0}) = ID:{1}", word, id)
                  If lexEnum.MoveTo(word) = False Then
                     Throw New ArgumentException("Could not position to " + word)
                  End If
                  sw.WriteLine("  MoveTo({0})  = ID:{1}", word, lexEnum.Index)
                  sw.Flush()
                  
               End If
            
            
            Else
               Throw New ArgumentException(inputFile + " does not exist.")
            End If
         Catch 
	   Throw
         Finally
            If Not (lex Is Nothing) Then
               lex.Dispose()
            End If
         End Try 
      End Sub 'DisplayWords
   End Class 'LexiconExamples
    _ 
   
End Namespace 'Microsoft.NaturalLanguage.Samples 

File: AllSamples/Samples/Samples_Source/natural language/lex/csharp/lex.cs
//---------------------------------------------------------------------
//  This file is part of the Microsoft .NET Framework SDK Code Samples.
// 
//  Copyright (C) Microsoft Corporation.  All rights reserved.
// 
//This source code is intended only as a supplement to Microsoft
//Development Tools and/or on-line documentation.  See these other
//materials for detailed information regarding Microsoft code samples.
// 
//THIS CODE AND INFORMATION ARE PROVIDED AS IS WITHOUT WARRANTY OF ANY
//KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
//IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
//PARTICULAR PURPOSE.
//---------------------------------------------------------------------
using System;
using System.IO;
using System.Collections;
using System.NaturalLanguage;
using System.Text;
using System.Globalization;

namespace Microsoft.NaturalLanguage.Samples
{
    class LexiconConsole
    {
        /// <summary>
        /// The main entry point for the Console.
        /// </summary>
        static int Main (string[] args)
        {
            LexiconExamples lexSample = null;
            LexiconCmdOpts cmdOpts = null;
            StreamWriter outputWriter = null;

            try
            {
                //Process commandline arguments
                cmdOpts = new LexiconCmdOpts();
                if (!cmdOpts.ProcessOpts(args))
                {
                    throw new ArgumentException("Parsing commandline arguments failed.");
                }

                lexSample = new LexiconExamples();
                if (cmdOpts.OutputFile != null)
                {
                    outputWriter = new StreamWriter(cmdOpts.OutputFile);
                }
                else
                {
                    outputWriter = new StreamWriter(Console.OpenStandardOutput());
                }

                lexSample.DisplayWords(outputWriter, cmdOpts.InputFile, cmdOpts.Word, cmdOpts.MaxWordsToDisplay);
                return 0;
            } 
            catch (System.NullReferenceException nre) 
            {
                if(cmdOpts.Verbose)
                {
                    Console.Error.WriteLine("Message: {0}", nre.Message);
                    Console.Error.WriteLine("Stack Trace: {0}", nre.StackTrace);
                }
                return 1;	
            } 
            catch (System.ArgumentException ae)
            {
                if (cmdOpts.Verbose)
                {
                    Console.Error.WriteLine("Message: {0}", ae.Message);
                    Console.Error.WriteLine("Stack Trace: {0}", ae.StackTrace);
                }

                return 1;
            }
            finally
            {
                if (outputWriter != null) outputWriter.Close();
            }
        }
    }

    public class LexiconExamples 
    {
        public LexiconExamples()
        {
        }

        protected void DisplayWordProps(StreamWriter sw, IDictionary props, string prefix)
        {
            foreach (DictionaryEntry pair in props)
            {
                sw.Write("{0}name: {1}", prefix, pair.Key.ToString());

                IDictionary subProps = null;
                ICollection subList = null;

                if (pair.Value is IDictionary)
                {
                    subProps = (IDictionary) pair.Value;
                    sw.WriteLine();
                    DisplayWordProps(sw, subProps, prefix + "  ");
                }
                else if (pair.Value is ICollection)
                {
                    subList = (ICollection) pair.Value;
                    string prefix2 = prefix + "\t";

                    sw.WriteLine();
                    foreach (object member in subList)
                    {
                        sw.WriteLine("{0} value: {1}", prefix2, member.ToString());
                    }
                }
                else
                {
                    sw.WriteLine(" value: {0}", pair.Value.ToString());
                }
            }

            sw.Flush();
        }

        public void DisplayWords (StreamWriter sw, string inputFile, string word, int maxWordsToDisplay)
        {
                LexiconEntryEnumerator lexEnum = null;
                Lexicon lex = null;

            if ((sw == null) || (inputFile == null) || (word == null) || (maxWordsToDisplay <= 0)) return;

            try
            {
                    if (File.Exists (inputFile))
                    {
                            //Create lexicon based on the lexicon file specified in the commandline
                            lex = new Lexicon (inputFile);
                            
                            //Get a LexiconEntryEnumerator
                            lexEnum = (LexiconEntryEnumerator)lex.GetEnumerator();

                            //Move to the target word if it is specified in the commandline
                            if (word.Length != 0)
                            {
                                    int id = lex.IndexOf (word);
                                    sw.WriteLine ("  IndexOf({0}) = ID:{1}", word, id);
                                    if (lexEnum.MoveTo(word) == false)
                                    {
                                            throw new ArgumentException("Could not position to " + word);
                                    }
                                    sw.WriteLine ("  MoveTo({0})  = ID:{1}", word, lexEnum.Index);
                                    sw.Flush();

                            }

                            
                    }
                    else
                    {
                            throw new ArgumentException (inputFile + " does not exist.");
                    }
            }
            catch (Exception)
            {
                    throw;
            }
            finally
            {
                if (lex != null)
                    lex.Dispose ();
            }

        }

    }

        /// <summary>
        /// The Lexicon Command Options class processes commandline arguments
        /// </summary>
        class LexiconCmdOpts : CmdOpts
        {
       
        }
}
